pkg/models/reading.go

```diff
@@ -21,32 +21,82 @@ import (
 /*
  * This file is for the Reading model in EdgeX
  * Holds data that was gathered from a device
- *
- *
+ */
+
+// ValueType indicates the type of value being passed back
+// from a ProtocolDriver instance.
+type ValueType int
```

Comment:
The indication of type is already present in a value descriptor. 
https://github.com/edgexfoundry/edgex-go/blob/4c54a00c141161af257cef1d7c161527c4d51ce7/pkg/models/value-descriptor.go#L34

Each reading is based on a value descriptor and we validate the reading accordingly upon ingestion into core-data. However it looks like for the SenML export, you need type information at export time in addition to the reading value and (understandably) don't want to go lookup the value descriptor for every reading to be exported. The SenML solution has a minimal typing in its reading that you're attempting to map to rather than using a default string property value.

Because of the change to the contract, I can't let this PR go through without discussion with the Device Services working group.

Reply:
Not sure I got your point... The problem I saw with export (or at least on how Device Services export data) is everything seems to be encoded as a string. However, Sen ML RFC explicitly states whether a value must be encoded as a boolean, as a string, as a float or as binary data in the json file. For instance, for the examples I am using (humidity, pressure and temperature), all of them are explicitly defined in the RFC as "float". 
So, if float is encoded using base64 as it is defined in pkg/models/commandvalue.go in device service (device sdk) and sent as a string, I am not able to know the type. I could not find another way to get it, but explicitly encoding, probably there are other ways to do it, but unfortunately I am a newbie when it comes to go lang and definitively don't know all the details of the Edge X Foundry architecture.
Of course I fully agree this is something to be discussed with Device Services.

Reply:
@tsconn23 @juanmagal 

The current approach where clients are required to use value descriptors in order introspect string reading values is a core concept of EdgeX. While the original Java code that was released to the community used a slightly different schema for value properties and the related value descriptors, the idea was still the same, the low-level device service protocol layer was responsible for translating native data types to a string value contained in a reading object. If a client needed detailed type information (i.e. like in your SenML export logic), the naive implementation would be to query the associated value descriptors on-demand. As the set of value descriptors in an EdgeX implementation don't change all that often, a client could easily cache all of the value descriptors locally and only query if/when a reading was encountered with an unknown descriptor.

With all that said, as the new type model is much simpler, we could in fact consider adding the type to the reading object, as it is only a single `int` field. The original Java code used a type scheme with multiple attributes, which I imagine was one of the factors in not replicating them all in each reading, but instead storing them in the value descriptor objects shared across a single EdgeX instance.

Reply:
Thanks! 
Besides the measure type, I am also including in the reading object the "unit" of the measure read which is a string, would you also consider including the unit as part of the reading?
@tsconn23 @tonyespy , should I assume the introduction of the type (and the unit) in the reading object will be accepted? Should I change something on my code?

Reply:
@tonyespy To your second point on adding the type to the reading object, I don't think @jpwhitemn is going to be onboard with that for Edinburgh. In discussions I've had with him, this falls under the auspices of the new application services which could provide the representation @juanmagal is looking for before exporting via SenML. I think Jim is going to resist changing the core model as is done in this PR.

Reply:
Thanks @tsconn23 
I am of course open to other alternatives if you think this change on the data model will not be accepted. IMHO I don't know if the best option is waiting till this new application service is ready, do you have any expected lead time? 
I could try to check the ValueDescriptor option even from performance point of view it might not be the best option. However, I am not sure how I can get a ValueDescriptor from the Reading object.

Reply:
@tsconn23 @tonyespy without knowing in depth new changes - how will an app in the cloud know the type of `Value` if all it recieves is a string? Under which type will it store it in the DB?

Regarding the timestamp - it can be always the current time in the moment of sending the message. It can be also set to `0`, as a relative offset of server time (time when message is recieved). Finally - `Time` field in SenML is optional - if it does not exist then server will probably attach timestamp of the receiving of the message.